Cycle Predictions

lacimonsrud opened the pull request:
Hey guys! Here is the draft PR for cycle predictions :)

Currently the frontend functionality isn't fully there yet, but on the backend predictions are being made based on the last fully documented cycle using the "Cycle" and "Cycle End" chips added to the Flow Accordion. 

Here are my organized thoughts about the full functionality of predictions, at least for now

- Prediction Enable/Disable
  - I added the option to turn cycle predictions on/off in the settings menu (doesn't do much yet, will enable/disable predictions on the calendar once that is implemented)
  - Settings menu has a button to view predicted cycle for debugging/testing purposes until the UI portion is fully worked out
  - On second thought, I'm curious if it would work better to just include predictions within the filter menu?  I originally put it in the settings menu to have an area to explain how the prediction algorithm works for user transparency. I can't decide if one would be best, or do you guys think it would even be necessary to have the settings tab for predictions if it can easily be turned on/off within the filters menu. My main point for this would be to make users aware of inaccuracy because the algorithm is very simple right now

- Algorithm
  - Currently, this feature works by taking the most recently documented cycle -and mirroring that cycle 3 weeks in the future, which is admittedly very simple. 
  - Planning for improvement to average out length & time between documented menstrual cycles for the finished simple version - going to finish UI then tackle this
  - The user HAS to mark their cycle start and end dates for the the algorithm to recognize menstrual cycles and therefore make a prediction. Is this going to be annoying/will users document these values? How will this work for people w/irregular cycles?
  - Looking for any suggestions or thoughts on how we can make cycle predictions the most accurate we can within our scope

- Calendar
  - Working on creating filters for both cycle start/end - should these only be one combined filter, or two?
  - How should we format the cycle start and cycle end date input options within the flow accordion? Is the way I did it fine, or would there be a better option? 
  - Random thought - but should we change the flow intensity options in the flow accordion to single selection chips to match the other accordions?
  - When should the prediction be removed from the calendar? When the current date reaches the predicted date? If the user has predicted cycles on, should we notify them when their cycle is starting? 

- Thoughts? Other Considerations?
Again, any ideas are greatly appreciated! 

- Video Demo
 
![ScreenRecording_05-12-202508-28-37_1-ezgif com-speed](https://github.com/user-attachments/assets/e992d6af-169e-450b-b003-045d0ff1237e)

lacimonsrud committed:
start/end options added for cycles

lacimonsrud committed:
Merge remote-tracking branch 'origin' into prediction

lacimonsrud committed:
added cycle prediction files

lacimonsrud committed:
fixed issues with start/end saving in db

lacimonsrud committed:
cycle prediction works on the most recently logged period, just need to implement to calendar

lacimonsrud committed:
tweaked some things, added debugger button

lacimonsrud committed:
changed cycle start/end btns to only show up once the user clicks on a flow_intensity

lacimonsrud committed:
format

arcanda3 commented:
Thank you so much for working on this feature Laci! 

Here's some of my thoughts:

- Prediction Enable/Disable
  - I like the prediction on/off button being in the settings since it allows space for the explanation of the algorithm and possible inaccuracy, which I think is important to make sure users see.
  - I think we could have the predicted dates as a filter, but I wonder if they could get included in the flow filter if prediction is on? Then just 1/3 of the filters on the calendar is handling cycle dates and users can use the 2 remaining filters for other data.
  - If the predicted dates are a separate filter, then I would say to still keep the prediction on/off in the settings, and have a pop up instructing the user to turn on predictions if they try to use the filter with predictions off. That way the user still has to go into settings to see the algorithm explanation and understand it may not be accurate.
- Algorithm
  - I can see the entering of the cycle start and cycle end dates getting annoying for users or user not documenting them accurately every cycle. Also, users most likely won't know their cycle end date on the day of, meaning they would have to remember to go back and enter that information later on.
  - When I was working on the markings for the flow chart, I was initially processing the data to find cycle periods by checking if the date previous to the current date had flow, and if no, it was marked as a cycle start. Then each next date was checked to find the last consecutive date with flow and marking that as the cycle end.
    - It wouldn't be perfect, but I wonder if something similar to this could be used to detect the start and end cycle dates on the backend instead of relying on the user?
- Calendar
  - Ideally, I think the filters for both cycle start/end should be one combined filter just to reduce the number of filters overall. However, I think it'd be fine either way.
  - I like the idea of making the flow options match the other accordions. I do like how they're currently stacked on top f each other as radio buttons, so maybe we could use the single selection chips and format them to be stacked in a column? Not sure how it would look on bigger screens like a tablet though if they take up the entire width for each chip.
  - The cycle start and end inputs look good how they are, and I like that they only show once a flow intensity other than none has been selected. Not sure how feasible this is, but I think it'd be nice if the cycle start/end inputs popped up right underneath the flow option that was selected. For example, if medium flow was selected, the inputs would pop up below medium and above heavy.
  - For displaying the prediction, I think it should stay on the calendar until the date has passed or the user logs flow for the predicted date. For example, if a user didn't log flow for a predicted date, it would still show as predicted for that entire day, but the next day, the prediction would be removed from the calendar.

lacimonsrud committed:
Merge branch 'main' into prediction

lacimonsrud committed:
cycle prediction filter working, bug with filter input

lacimonsrud committed:
format

lacimonsrud committed:
removed unecessary imports

lacimonsrud commented:
Update: Cycle Prediction is added as a filter option now! 
Here is what I'm now working on:
- Re-implementing the Settings accordion for Cycle Prediction (removed when figuring out filter)
- Changing Cycle Prediction from it's own filter to be included within the flow filter if cycle predictions are enabled in the settings menu
- Changing algorithm to not operate on flow start/end dates, but instead on continuous flow entries
- Removing predictions once the date has passed

arcanda3 commented:
> Definitely something that would drive me a little crazy and may even make me less inclined to use the app, so I think that I'll see what I can learn from your FlowChart component and try that!

The code that was removed in this commit https://github.com/adulbrich/ephira/commit/926926bf2ea0afbb11b53d0b1b8fad36ee7a10ce is all the logic for grouping together the consecutive flow dates. The deleted lines were also to render the curves I was attempting to add on the chart, but you should be able to grab the grouping logic and see if it will work for cycle predictions!

lacimonsrud committed:
cycle prediction toggling implemented

lacimonsrud committed:
format

lacimonsrud commented:
> The code that was removed in this commit [926926b]

Thank you! This is perfect :)

lacimonsrud committed:
changed cycle prediction to have it's own filter

lacimonsrud committed:
Cycle predictions now don't require stop/start, determined by latest logged cycle (3 days +)

lacimonsrud committed:
finishing touches..

lacimonsrud committed:
added cycle start/end to filter menu

lacimonsrud committed:
updated markeddates

lacimonsrud commented:
Predictions are now only based off of consecutive days logged (I thought that ~3 days would be a good length for consecutive days to count as a cycle?). I ended up making the cycle predictions filter seperate from the flow filter in order to allow it to have it's own spot in the filter legend. This filter option is removed when cycles are disabled. Also, the cycle predictions will exist on the calendar until the day that the prediction is for passes, then it will no longer be marked. 

Also added start/end cycle option to the filters menu.

I tweaked a few things after opening up this PR to be merged a couple of days ago as I felt it was pretty much complete, and now I think it's ready to be merged as long as there are no bugs or suggestions to be made. 

I did, however, notice that the radio buttons on the filter menu are being really buggy, but upon further inspection I think this was introducted in another PR? (Feel free to correct me if I'm wrong here). I couldn't trace back to exactly when, but with the filter menu having the current filters pop up on the top bar with the quick "X" option to disable them and also the radio buttons for turning filters on/off, this doesn't seem to be synced. I can investigate a bit more then open up a new issue if necessary.

arcanda3 commented:
Looks good to me and works well! One minor suggestion I have is possibly adding a brief note in the cycle predictions accordion on the settings screen that predicted dates will not show if they are in the past. When I first started testing I thought it wasn't working because I saw no predicted dates on the calendar, but then I realized it was because I didn't have recent enough flow logged for the predicted dates to be in the future. Adding a note about this would make sure users don't run into the same confusion.

arcanda3 commented:
> I did, however, notice that the radio buttons on the filter menu are being really buggy, but upon further inspection I think this was introducted in another PR? (Feel free to correct me if I'm wrong here). I couldn't trace back to exactly when, but with the filter menu having the current filters pop up on the top bar with the quick "X" option to disable them and also the radio buttons for turning filters on/off, this doesn't seem to be synced. I can investigate a bit more then open up a new issue if necessary.

I noticed this as well, and I'm also not sure when it was introduced. The same bugginess occurs with the switches in the Calendar Entries Visibility toggles, so I believe it's a general issue with the component and state. However, I think I figured out a fix for this, and I plan on making a PR that addresses this today.

steele836 committed:
Fixed two small typos and slight rewording

- predictionns -> predictions
- developmental determines -> developmental and determines
- cycle by consecutive -> cycle based on consecutive

lacimonsrud committed:
added more info to the settings page on cycle predictions

lacimonsrud committed:
broke up text sections within cycle predictions settings dialog

lacimonsrud committed:
format
